src/meds_torch/models/components/transformer_decoder.py

```diff
@@ -1,15 +1,286 @@
+        >>> new_time, continue_gen = update_generation_budget(
+        ...     cumulative_time=cumulative,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=2,
+        ...     get_next_token_time=get_time,
+        ...     budget=time_budget,
+        ... )
+        >>> new_time  # Should be 1
+        tensor([1., 1.])
+        >>> continue_gen  # Should continue since < min_time_len
+        True
+        >>> new_time, continue_gen = update_generation_budget(
+        ...     cumulative_time=new_time,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=3,
+        ...     get_next_token_time=get_time,
+        ...     budget=time_budget,
+        ... )
+        >>> continue_gen
+        False
+        >>> new_time # Should be incremented by 1
+        tensor([2., 2.])
+
+        >>> seq_len_budget = GenerationBudget.from_seq_len(10)  # Generate until max_seq_len >= 10
+        >>> # Test sequence length budget
+        >>> _, continue_gen = update_generation_budget(
+        ...     cumulative_time=cumulative,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=5,
+        ...     get_next_token_time=None,
+        ...     budget=seq_len_budget,
+        ... )
+        >>> continue_gen  # Should continue
+        True
+        >>> seq_len_budget = GenerationBudget.from_seq_len(5)  # Generate until max_seq_len >= 10
+        >>> # Test sequence length budget
+        >>> _, continue_gen = update_generation_budget(
+        ...     cumulative_time=cumulative,
+        ...     current_sample=current_sample,
+        ...     num_generated_tokens=5,
+        ...     get_next_token_time=None,
+        ...     budget=seq_len_budget,
+        ... )
+        >>> continue_gen # Should stop since > max_seq_len
+        False
+    """
+    # Get budget constraints
+    min_time_len = budget.min_time_len
+    is_time_budget = exists(min_time_len)
+    if is_time_budget:
+        # Get time prediction for the new token
+        with torch.no_grad():
+            if get_next_token_time is None:
+                raise ValueError("`get_next_token_time` is required when generating using the time budget")
+            pred_time = get_next_token_time(current_sample)
+            cumulative_time = cumulative_time + pred_time.squeeze(-1)
+
+        continue_generation = (cumulative_time < min_time_len).any().item()
+    else:
+        continue_generation = num_generated_tokens < budget.max_seq_len
+
+    return cumulative_time, continue_generation
+
+
 class TransformerDecoderModel(torch.nn.Module, Module):
-    """Wrapper of Decoder Transformer for use in MEDS with triplet token embeddings."""
+    """Wrapper of Decoder Transformer for use in MEDS with triplet token embeddings.
+
+    This model handles both forward passes and generation with different budget types.
+
+    Examples:
+        >>> # Setup mock configuration
```

Comment:
I'd put some/all of this setup in a function so that these doctests are easier to read.